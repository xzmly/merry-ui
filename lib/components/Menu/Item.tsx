import * as React from 'react';
import { useContext } from "react"
import classes from '../../helpers/classes';
import MenuContext,{RestDataType,MenuContextProps}from "./Context";

export interface MenuItemProps extends React.HTMLAttributes<HTMLElement> {
  _key: string
  restData?: RestDataType
  disabled?: boolean
  paddingLeft?:number
}

const Item: React.FC<MenuItemProps> =
    props => {

      const {
        onSelect,
        selectedKeys,
        defaultSelectedKeys
      } = useContext<MenuContextProps>(MenuContext);

      const {
        className,
        children,
        _key,
        restData,
        disabled,
        paddingLeft
      } = props;

      const names:Array<string> = selectedKeys || defaultSelectedKeys || [];

      const onClick = (e:React.MouseEvent<HTMLElement>):void => {
        if (disabled) return;
        onSelect && onSelect({
          keys: [_key],
          event: e,
          restData
        })
      };

      return (
          <li className={classes(
              className,
              'menu-item',
              names.includes(_key) ? 'active' : "",
              disabled ? "disabled" : ""
          )}
              onClick={onClick}
              style={{paddingLeft: `${paddingLeft}px`}}
          >
            {children}
          </li>
      )
    };

export default Item