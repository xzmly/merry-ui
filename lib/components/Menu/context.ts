import * as React from "react"

export type RestDataType = {
  item: object
}

export type OnChangeType = {
  names: Array<string>
  event:React.MouseEvent
  restData?: RestDataType
};

export interface MenuContextProps {
  openKeys?: Array<string>
  onSubMenuChange?: (params: OnChangeType) => void
  defaultOpenKeys?: Array<string>

  onSelect?: (params: OnChangeType) => void
  defaultSelectedKeys?: Array<string>
  selectedKeys?: Array<string>
}

const MenuContext = React.createContext<MenuContextProps>({
  openKeys: [],
  onSubMenuChange: (params: OnChangeType) => null,
  defaultOpenKeys: [],
  onSelect: (params: OnChangeType) => null,
  defaultSelectedKeys: [],
  selectedKeys: []
});

export default MenuContext