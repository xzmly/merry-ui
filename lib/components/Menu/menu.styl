@import '../../helpers/stylus.helper.styl';

menu = class-prefix + '-menu'
sub-menu = class-prefix + '-sub-menu'
menu-item = class-prefix + '-menu-item'
font-color = rgba(0,0,0,0.75)

common-hover()
  &:hover
    color primary-color
    transition all .2s

.{sub-menu}-content,.{menu}
  list-style-type none
  padding-inline-start 0


.{menu}
  min-height 80px

.{sub-menu}
  color rgba(0,0,0,0.75)
  font-size 14px
  &.{class-prefix}-active
    > .{sub-menu}-title
      color primary-color
      > .{sub-menu}-arrow
        transform rotate(-180deg)
        color font-color
  .{sub-menu}-title
    display flex
    justify-content space-between
    align-items center
    height 40px
    cursor pointer
    padding-right: 16px;
    common-hover()
    .{sub-menu}-arrow
      font-size 12px
      transition all .2s
  .{sub-menu}-content
    padding: 0

.{menu-item}
  common-hover()
  cursor pointer
  height 40px
  display flex
  justify-content space-between
  align-items center
  padding-right: 16px;
  &.{class-prefix}-active
    background rgba(21, 171, 255, 0.24)
    color primary-color
    border-right 3px solid primary-color
