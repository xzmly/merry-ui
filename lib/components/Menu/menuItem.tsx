import * as React from 'react';
import classes from '../../helpers/classes';

export interface MenuItemProps extends React.HTMLAttributes<HTMLElement> {
}

const MenuItem: React.FC<MenuItemProps> =
    props => {

      const { className,children } = props;

      return (
          <div className={classes(className,'menu-item')}>
            {children}
          </div>
      )
    };

export default MenuItem