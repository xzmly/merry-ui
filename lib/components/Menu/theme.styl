
theme(background,li-font-color,item-active-background,item-border)
  background background
  li
    color li-font-color
  .{sub-menu}
    &.{class-prefix}-active
      > .{sub-menu}-title
        color rgba(255,255,255,1)
        > .{sub-menu}-arrow
          color rgba(255,255,255,1)
    .{sub-menu}-title
      common-hover(rgba(255,255,255,1))
  .{item-group} > .{item-group}-title
    color rgba(255,255,255,0.5)
  .{menu-item}
    common-hover(rgba(255,255,255,1))
    &.{class-prefix}-active
      color rgba(255,255,255,1)
      background item-active-background
      border-right 3px solid item-border