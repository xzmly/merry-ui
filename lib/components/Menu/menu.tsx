import * as React from 'react';
import classes from '../../helpers/classes';
import './menu.styl';
import {HTMLAttributes} from "react";

interface MenuProps extends HTMLAttributes<HTMLDivElement> {

}


const Menu: React.FC<MenuProps> =
        props => {

  const { className,children } = props;
  return (
      <div className={classes(className,'menu')}>
        {children}
      </div>
  )
};

export default Menu