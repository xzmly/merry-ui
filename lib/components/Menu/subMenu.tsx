import * as React from 'react';
import { useContext } from "react"
import classes from '../../helpers/classes';
import Icon from "../Icon/icon";
import {ChildrenType} from "./menu"
import MenuContext,{RestDataType}from "./context";

export interface SubMenuProps {
  title?: React.ReactNode
  className?: string
  children?: ChildrenType | Array<ChildrenType>
  name: string
  style?: React.CSSProperties
  restData?: RestDataType
  paddingLeft?: number
}

const SubMenu: React.FC<SubMenuProps> =
    props => {

      const {
        openNames,
        onSubMenuChange,
        defaultOpenNames,
      } = useContext(MenuContext);

      const {
        className,
        children,
        title,
        name,
        restData,
        paddingLeft
      } = props;

      const defaultNamesOrNames:Array<string> = openNames || defaultOpenNames || [];

      const visible:boolean = defaultNamesOrNames.includes(name);

      const toggleVisible = (e:React.MouseEvent<HTMLElement>):void => {
        const newKeys = visible ?
            defaultNamesOrNames.filter(v => v !== name) :
            [name,...defaultNamesOrNames];

        onSubMenuChange && onSubMenuChange({
          names: newKeys,
          event: e,
          restData
        });
      };

      const childrenClass = (postfix:string,...names: Array<string>):string =>
          classes("",`sub-menu-${postfix}`,...names);


      return (
          <li className={classes(className,'sub-menu',visible ? 'active':'')}>
            <div className={childrenClass('title')}
                 onClick={toggleVisible}
                 style={{paddingLeft: `${paddingLeft}px`}}
            >
              {title}
              <Icon name={'arrow'}
                    className={childrenClass('arrow')}/>
            </div>
            {visible &&
            <ul className={childrenClass('content')}>
              {children}
            </ul>}
          </li>
      )
    };

export default SubMenu