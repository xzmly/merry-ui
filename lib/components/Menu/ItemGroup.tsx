import * as React from 'react';
import classes from '../../helpers/classes';
import {MenuItemProps} from "./Item";
import {SubMenuProps} from "./subMenu";

type ChildrenType = React.ReactElement<MenuItemProps> | React.ReactElement<SubMenuProps>

export interface ItemGroupProps {
  children?: ChildrenType | Array<ChildrenType>
  className?: string
  title: React.ReactNode
}

const ItemGroup: React.FC<ItemGroupProps> =
        props => {

  const { children,className,title } = props;

  return (
      <li className={classes(className,'menu-item-group')}>
        <div className={classes(undefined,'menu-item-group-title')}
        >
          {title}
        </div>
        <ul>
          {children}
        </ul>
      </li>
  )
};

export default ItemGroup