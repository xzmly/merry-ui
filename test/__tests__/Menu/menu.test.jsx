import Menu from '../../../lib/components/Menu/menu'
import {mount} from "enzyme/build/index";
import React from 'react'
import renderer from "react-test-renderer";

const { MenuItem,SubMenu,ItemGroup } = Menu;

const example = renderer.create(
    <Menu>
      <MenuItem key={'1'} _key={'1'}>option1</MenuItem>
      <SubMenu _key={'2'} key={'2'}>
        <MenuItem key={'3'} _key={'3'}>option2</MenuItem>
        <SubMenu _key={'8'} key={'8'}>
          <MenuItem key={'9'} _key={'9'}>option2</MenuItem>
        </SubMenu>
      </SubMenu>
      <ItemGroup title={'通用'} key={'5'}>
        <MenuItem key={'4'} _key={'4'}>option3</MenuItem>
        <SubMenu _key={'6'} key={'6'}>
          <MenuItem key={'7'} _key={'7'}>option2</MenuItem>
        </SubMenu>
      </ItemGroup>
    </Menu>
);

describe('Menu',()=> {
  it('menu 是个 ul',()=>{
    const json = renderer.create(<Menu/>).toJSON();
    expect(json).toMatchSnapshot()
  })
  it('menu 接受 className', () => {
    const c = mount(<Menu className='menu'/>);
    expect(c.find('ul').hasClass('menu')).toBe(true)
  });
  it('menu 接受 theme', () => {
    const c = mount(<Menu theme='dark'/>);
    expect(c.find('ul').hasClass('merry-menu-theme-dark')).toBe(true)
    c.setProps({ theme: 'blue' });
    expect(c.find('ul').hasClass('merry-menu-theme-blue')).toBe(true)
    c.setProps({ theme: 'default' });
    expect(c.find('ul').hasClass('merry-menu-theme-default')).toBe(false)
  });
  it('menu 接受 style', () => {
    const c = mount(<Menu style={{margin: "10px"}}/>);
    const style = c.props().style
    expect(style.hasOwnProperty('margin')).toBe(true)
  });
  it('menu 的 children', () => {
    expect(example).toMatchSnapshot()
  });
  it('')
})