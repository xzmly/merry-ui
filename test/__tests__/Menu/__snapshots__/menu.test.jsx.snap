// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu menu 是个 ul 1`] = `
<ul
  className="merry-menu"
/>
`;

exports[`Menu menu 的 children 1`] = `
<ul
  className="merry-menu"
>
  <li
    className="merry-menu-item"
    onClick={[Function]}
    style={
      Object {
        "paddingLeft": "20px",
      }
    }
  >
    option1
  </li>
  <li
    className="merry-sub-menu"
  >
    <div
      className="merry-sub-menu-title"
      onClick={[Function]}
      style={
        Object {
          "paddingLeft": "20px",
        }
      }
    >
      <svg
        className="merry-icon merry-sub-menu-arrow"
      >
        <use
          xlinkHref="#arrow"
        />
      </svg>
    </div>
  </li>
  <li
    className="merry-menu-item-group"
  >
    <div
      className="merry-menu-item-group-title"
    >
      通用
    </div>
    <ul>
      <li
        className="merry-menu-item"
        onClick={[Function]}
        style={
          Object {
            "paddingLeft": "20px",
          }
        }
      >
        option3
      </li>
      <li
        className="merry-sub-menu"
      >
        <div
          className="merry-sub-menu-title"
          onClick={[Function]}
          style={
            Object {
              "paddingLeft": "20px",
            }
          }
        >
          <svg
            className="merry-icon merry-sub-menu-arrow"
          >
            <use
              xlinkHref="#arrow"
            />
          </svg>
        </div>
      </li>
    </ul>
  </li>
</ul>
`;
